### PR TITLE
Adds facts importing background task

### PR DESCRIPTION
**Pull request by ares**

### CODE REVIEW

File: app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+module ForemanBackground::API::HostsControllerExtensions
+  extend ActiveSupport::Concern
+  included do
+    alias_method_chain :facts, :background
+  end
+
+  def facts_with_background
+    if ReportWorker.redis_available?
+      facts = params[:facts].to_json
+      FactWorker.perform_async(detect_host_type.to_s, params[:name], params[:facts], params[:certname])
```
Comment (ohadlevy) [inline app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb:10]: I'm not sure if we should always push facts via bg, as the ENC output might depend on it.
I think it makes sense when we do not use ENC(either disabled or not using puppet?)

Comment (iNecas) [inline app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb:10]: Or, we could also teach facts uploader to wait for the async operation to finish (polling interval of 10 seconds might work quite well without overloading the web server)

**Comment by ares** [comment]
Currently every importer can specify whether it supports background processing of facts or not. I tested with both puppet and chef. foreman_chef gem supports this from version 0.0.2. I think we should release this plugin officially so users can use it.

**Comment by ares** [comment]
ping @ohadlevy

File: app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
+module ForemanBackground::API::HostsControllerExtensions
+  extend ActiveSupport::Concern
+  included do
+    alias_method_chain :facts, :background
+  end
+
+  def facts_with_background
+    if FactWorker.redis_available? && facts_import_supports_background?
+      facts = params[:facts].to_json
```
Comment (ohadlevy) [inline app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb:9]: why move it back to json? it was already received as json and parsed by rails?

Comment (ohadlevy) [inline app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb:9]: hmm..  are we using it at all? there is also a facts method which is a bit confusing.

File: app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      logger.warn 'REDIS is not available , blocking until report processing is done.' unless FactWorker.redis_available?
+      logger.warn 'Import type does not support background processing, blocking until report processing is done.' unless facts_import_supports_background?
+      facts_without_background
+    end
+  rescue ::Foreman::Exception => e
+    render :json => {'message'=>e.to_s}, :status => :unprocessable_entity
+  end
+
+  def facts_import_supports_background?
+    @facts_import_supports_background ||= begin
```
Comment (ohadlevy) [inline app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb:22]: should the begin/end block here could be extracted to a method?

File: app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    alias_method_chain :facts, :background
+  end
+
+  def facts_with_background
+    if FactWorker.redis_available? && facts_import_supports_background?
+      facts = params[:facts].to_json
+      FactWorker.perform_async(detect_host_type.to_s, params[:name], params[:facts], params[:certname])
+      head :status => 202
+    else
+      logger.warn 'REDIS is not available , blocking until report processing is done.' unless FactWorker.redis_available?
```
Comment (ohadlevy) [inline app/controllers/concerns/foreman_background/api/hosts_controller_extensions.rb:13]: nitpick: whitespace between ',' char

**Comment by ohadlevy** [comment]
there is some overlap with https://github.com/ohadlevy/foreman-background/commit/3872a6890c0bba8cab3f768b710bd7c5322a3055 could you please have a look and merge the similar things?

thanks!

**Comment by ares** [comment]
I guess I can close this one, we have a different technology after 3 years :-)

**Pull request closed without merging**